tests/storage/blocksd_test.py

```diff
@@ -1244,6 +1251,117 @@ def test_create_volume_with_new_bitmap(
+def test_merge_with_bitmap(
+        tmp_storage, tmp_repo, fake_access, fake_rescan, tmp_db, fake_task,
+        fake_sanlock, monkeypatch):
+    sd_id = str(uuid.uuid4())
+
+    dev = tmp_storage.create_device(10 * GiB)
+    lvm.createVG(sd_id, [dev], blockSD.STORAGE_UNREADY_DOMAIN_TAG, 128)
+    vg = lvm.getVG(sd_id)
+
+    dom = blockSD.BlockStorageDomain.create(
+        sdUUID=sd_id,
+        domainName="domain",
+        domClass=sd.DATA_DOMAIN,
+        vgUUID=vg.uuid,
+        version=4,
+        storageType=sd.ISCSI_DOMAIN)
+
+    sdCache.knownSDs[sd_id] = blockSD.findDomain
+    sdCache.manuallyAddDomain(dom)
+
+    dom.refresh()
+    dom.attach(tmp_repo.pool_id)
+
+    img_id = str(uuid.uuid4())
+    base_vol_uuid = str(uuid.uuid4())
+    top_vol_uuid = str(uuid.uuid4())
+
+    # Create base volume.
+    dom.createVolume(
+        imgUUID=img_id,
+        capacity=1 * GiB,
+        volFormat=sc.COW_FORMAT,
+        preallocate=sc.SPARSE_VOL,
+        diskType='DATA',
+        volUUID=base_vol_uuid,
+        desc="Test volume",
+        srcImgUUID=sc.BLANK_UUID,
+        srcVolUUID=sc.BLANK_UUID,
+        bitmap="new-bitmap")
+
+    base_vol = dom.produceVolume(img_id, base_vol_uuid)
+
+    # Create top volume with bitmaps.
+    dom.createVolume(
+        imgUUID=img_id,
+        capacity=1 * GiB,
+        volFormat=sc.COW_FORMAT,
+        preallocate=sc.SPARSE_VOL,
+        diskType='DATA',
+        volUUID=top_vol_uuid,
+        desc="Test top volume",
+        srcImgUUID=base_vol.imgUUID,
+        srcVolUUID=base_vol.volUUID,
+        add_bitmaps=True)
+
+    top_vol = dom.produceVolume(img_id, top_vol_uuid)
+
+    # Create leaf volume.
+    dom.createVolume(
+        imgUUID=img_id,
+        capacity=1 * GiB,
+        volFormat=sc.COW_FORMAT,
+        preallocate=sc.SPARSE_VOL,
+        diskType='DATA',
+        volUUID=str(uuid.uuid4()),
+        desc="Test top volume",
+        srcImgUUID=top_vol.imgUUID,
+        srcVolUUID=top_vol.volUUID,)
+
+    host_id = 0
+    subchain_info = dict(sd_id=base_vol.sdUUID,
+                         img_id=base_vol.imgUUID,
+                         base_id=base_vol.volUUID,
+                         top_id=top_vol.volUUID,
+                         base_generation=0)
+    subchain = merge.SubchainInfo(subchain_info, host_id)
+
+    # Nearly fill the volume with data.
+    top_vol.prepare()
+    data_size = GiB - 5 * MiB
+    qemuio.write_pattern(
+        top_vol.getVolumePath(),
+        qemuimg.FORMAT.QCOW2,
+        offset=0,
+        len=data_size)
+    top_vol.teardown(sd_id, top_vol_uuid)
+
+    # Add bitmaps to the base volume after creating top volume.
+    base_vol.prepare()
+    for i in range(100):
+        qemuimg.bitmap_add(base_vol.getVolumePath(), f"bitmap{i}").run()
+    base_vol.teardown(sd_id, base_vol_uuid)
+
+    with monkeypatch.context() as mc:
+        mc.setattr(guarded, 'context', fake_guarded_context())
```

Comment:
Yes we need it. If we don't, then the `prepare()` operation timeouts without acquiring the lock.

So we need to call `dom.acquireHostId`, but then, even if we release it, the `test_spm_lifecycle` was failing somewhere in the sanlock code.

I saw this mock in the `merge_test` module, and since the guarded context is not important for this test, I imported it here too.

Reply:
Maybe add a TODO to check why this is needed.